src/qrgenerator.ts

```diff
@@ -0,0 +1,60 @@
+import { createHmac } from "node:crypto";
+
+type QrCache = {
+  startTime: number;
+  qrStartToken: string;
+  qrStartSecret: string;
+};
+
+interface QRGenerateOptions {
+  /** max cycles */
+  maxCycles?: number;
+  /** in seconds  */
+  timeout?: number;
+}
+
+export class QrGenerator {
+  cache = new Map<string, QrCache>();
+
+  constructor(qrStartSecret: string, qrStartToken: string, orderRef: string) {
+    const now = Date.now();
+    const qrStore = { startTime: now, qrStartSecret, qrStartToken };
+    this.cache.set(orderRef, qrStore);
+    return this;
+  }
+
+  /**
+   * Generator yielding a new value for the qrcode within
+   * the specified limits.
+   **/
+  *nextQr(
+    orderRef: string,
+    { maxCycles, timeout }: QRGenerateOptions = { timeout: 60 },
+  ) {
+    const qr = this.cache.get(orderRef);
+    if (!qr) {
+      return;
+    }
+    for (let i = 0; i >= 0; i++) {
+      const secondsSinceStart = parseInt(
+        `${(Date.now() - qr.startTime) / 1000}`,
+        10,
+      );
```

Comment:
Maybe simpler approach is to do `Math.floor((Date.now() - qr.startTime) / 1000)`?

Reply:
Or even better, throttle + memoize this generation properly such that we will only get a new code every second that has passed 🤔?